modules/assim.sequential/R/SDA_OBS_Assembler.R

```diff
@@ -0,0 +1,175 @@
+#' Assembler for preparing obs.mean and obs.cov for the SDA workflow
+#'
+#' @param settings_dir the path of settings.xml object.
+#' @param Var Variable name, currently support: SMP, AGB, and LAI.
+#' @param OutDir the path to store obs.mean and obs.cov
+#' @param Obs_Prep if your settings object doesn't contain Obs_Prep, you can import it separately (details see L17-18).
+#' @param skip_buffer flag to skip calculating min var based on buffer area for agb data.
+#'
+#' @return list of obs.mean and obs.cov
+#' @export
+#' @author Dongchen Zhang
+#'
+#' @examples
+#' \dontrun{
+#' settings_dir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/NEON42/SDA/pecan.xml"
+#' OutDir <- "/projectnb/dietzelab/dongchen/All_NEON_SDA/test_OBS"
+#' Var <- c("SMP", "LAI", "AGB")
+#' OBS <- SDA_OBS_Assembler(settings_dir, Var, OutDir)
+#' }
+
+
+SDA_OBS_Assembler <- function(settings_dir, Var, OutDir, Obs_Prep = NULL, skip_buffer = TRUE){
+  #export special operator
+  `%>%` <- magrittr::`%>%` 
+  `%m+%` <- as.function(lubridate::`%m+%`)
+  
+  #read settings
+  settings <- PEcAn.settings::read.settings(settings_dir)
+  
+  # I prefer to include Obs_Prep info in the settings object, that will make life easier!
+  # if you don't like it, you can provide the Obs_Prep object 
+  # followed strictly by L76 - L82 in "/pecan/modules/assim.sequential/inst/MultiSite-Exs/SDA/Create_Multi_settings.R"
+  if(!is.null(settings$state.data.assimilation$Obs_Prep)){
+    Obs_Prep <- settings$state.data.assimilation$Obs_Prep
+  }else if(is.null(Obs_Prep)){
+    PEcAn.logger::logger.info("Please provide info of Obs_Prep object!")
+    return(0)
+  }
+  
+  #prepare site_info offline, because we need to submit this to GEO, which doesn't support Bety connection.
+  
+  Site_Info <- list(site_id = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('id') %>% unlist() %>% as.character(),
+                    lat = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lat') %>% unlist() %>% as.numeric(),
+                    lon = settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lon') %>% unlist() %>% as.numeric(),
+                    site_name = rep("name", length(settings %>% purrr::map(~.x[['run']] ) %>% purrr::map('site') %>% purrr::map('lat') %>% unlist() %>% as.numeric())))
+  
+  #collect time points
+  #we need to know which var we want to proceed 
+  #cause for every variable we assigned the same time_step object, we only need to grab it from any of what we have here.
+  #here we grab the first var to calculate the time step.
+  var_first <- Var[1]
+  for (i in 1:length(Obs_Prep)) {
+    if(!is.character(try(Obs_Prep[[i]]$Var == var_first, silent = T)))Time_Step <- Obs_Prep[[i]]$Time_Step
+  }
+  #time operations
+  if(Time_Step$unit == "year"){
```

Comment:
This is fine for now but it seems like there should be a more generic way of doing this. For now, you definitely need to document in the pecan.xml that "year" and "day" are the only allowable units. Also, you should add a 3rd case to catch the error of someone not specifying the timestep units correctly (e.g. If I put in "month" then your code will happily continue without time_points being defined).

Reply:
added if control for detecting the input timestep unit.